### ISSUE TITLE
Selective hide of blocks in the palette

### ISSUE THREAD

**Issue by sfederici** [ScratchAddons/ScratchAddons#4125]
### Is your request related to a problem? Please describe.

Many teachers see their students often lost when -while starting to learn Scratch- they are faced to the many blocks available in the Scratch palette. Teachers would like to progressively show the Scratch blocks to their students, so that they are not frustrated by having to go through the long list of blocks. Solving this problem would make Scratch even more fit as a coding teaching tool.

### Describe the solution you'd like

Allow the users to selectively hide the unwanted blocks. Save the list of hidden blocks in the Scratchaddons settings (so that teachers can export them and let their students import them)

### Describe alternatives you've considered

None

### Additional context

This is very convenient and already available in coding sites like code.org or Scratch clones such as Snap (https://snap.berkeley.edu/snap/snap.html#, "Hide blocks" option in "File" menu). 

<img width="510" alt="Schermata 2022-01-25 alle 09 42 29" src="https://user-images.githubusercontent.com/2651868/150942188-f3ec5ae2-d6fe-4727-8cc5-45ce144b7a6c.png">

<img width="712" alt="Schermata 2022-01-25 alle 09 44 24" src="https://user-images.githubusercontent.com/2651868/150942661-f2f282c4-9645-4b56-bdb0-dc4bdbef8614.png">

**Comment by Mbrick2** [comment]
I feel like apple will close this

**Comment by sfederici** [comment]
> I feel like apple will close this

Apple? Nickname of one of the developers?

**Comment by sfederici** [comment]
> > I feel like apple will close this
> 
> Apple? Nickname of one of the developers?

Anyway, why do you fill that?

**Comment by apple502j** [comment]
Yes, I'm apple, and I'm voting no to this.

Scratch's main goal is to promote creative learning; the students should be able to code their own program, using whatever blocks they want, instead of doing something ordered by a teacher. While I understand this is not the goal of most of the traditional school programs, I do not feel like such features should be included in Scratch Addons. (Plus, I think Scratch has too few blocks.)

**Comment by sfederici** [comment]
> Yes, I'm apple, and I'm voting no to this.
> 
> Scratch's main goal is to promote creative learning; the students should be able to code their own program, using whatever blocks they want, instead of doing something ordered by a teacher. While I understand this is not the goal of most of the traditional school programs, I do not feel like such features should be included in Scratch Addons. (Plus, I think Scratch has too few blocks.)

As a teacher (I've been teaching with Scratch since 2008 and I have introduced block's hiding in Snap in my own mod before it got added to the official tool due to a huge number of requests from the teachers, despite the initial refusal by the tool designers) I can't agree. And I don't see why to refuse an addon very useful from an educational viewpoint. Kids of course have not to use this add-on by themselves. But kids that are frustrated by the large Scratch palette will take advantage of this possibility offered to them by their teachers, without dropping learning it. So, if someone wants to give it a try, I would encourage them.

**Comment by sfederici** [comment]
> (Plus, I think Scratch has too few blocks.)

If you want many many blocks give a look at Snap ;-)

**Comment by WorldLanguages** [comment]
I don't think this should be completely rejected, but this is still a weird suggestion. If someone creates this addon, it will be accepted, but how it gets tagged should be discussed (danger tag? we don't want users telling us they don't find some blocks)
Also, we can't list every existing block in the settings page, so the list of hidden blocks to hide would need to be configured from the editor itself, and we don't really have a good "storage API" for that yet. So at the moment, this addon would be pretty useless, because importing settings json (which is probably what @sfederici is doing) wouldn't import the list of blocks to hide.

**Comment by Mbrick2** [comment]
What will happen if a project has one of these blocks. Then what

**Comment by CST1229** [comment]
> What will happen if a project has one of these blocks. Then what

I think just have that block appear as normal but not in the palette (like how Snap<i>!</i> does it).

**Comment by sfederici** [comment]
> we don't want users telling us they don't find some blocks

This can happen in two cases: you hide the block yourself (and you won't see it, as you desire) or someone asks you to import a set of settings, and you -if they don't tell you what will happen- you can expect something happening to your IDE. In both cases you can go to settings and to revert the setting.

> we can't list every existing block in the settings page

When organized in categories, Scratch blocks won't take so much space. Add-on settings can be collapsed, so it shouldn't be a huge problem

> importing settings json (which is probably what @sfederici is doing) wouldn't import the list of blocks to hide.

This if it is implemented in the IDE like in Snap, right?

**Comment by sfederici** [comment]
> I think just have that block appear as normal but not in the palette (like how Snap! does it).

Yes

**Comment by mxmou** [comment]
> > we can't list every existing block in the settings page
> 
> What's preventing us from doing so?

It would be very hard to find the block you want to hide. I think a way to hide a setting from the setting page and an `addon.settings.set` API (#1950) would be the best way to implement this.

**Comment by sfederici** [comment]
> The amount of space it would take up. !

I understand that when the setting is open it is going to take a large space. But, if you are interested in it, you don't care about how much space it is going to take. And, if you are not interested, you can leave the add-on collapsed (or collapse it back when you have had a look at it and found that it doesn't do anything good for you)

**Comment by sfederici** [comment]
BTW, it would be good to explicitly group the blocks by category, so that you can find the relevant blocks faster

**Comment by Chiroyce1** [comment]
As a teen who got into the world of programming via Scratch - I see this as a disadvantage for the students. Once they understand the basics of programming they should be allowed to use a block as per their choice, you can guide them and tell them more efficient methods, but hiding certain blocks would only hinder their creativity (as Apple said). And the title of this extension is *Scratch **Addons*** which is certainly meant to increase the possibilities with Scratch. 

I understand that as a teacher you'd like them to learn using only certain blocks, so that they learn more, but I feel like the majority of users who use Scratch Addons wouldn't find this interesting.

**Comment by sfederici** [comment]
> once they understand the basics of programming they should be allowed to use a block as per their choice, you can guide them and tell them more efficient methods, but hiding certain blocks would only hinder their creativity (as Apple said)

Sure. The teachers are not going to hide some Scratch blocks to their students forever. They are just allowing them to gradually apply a minimal effort, releasing new blocks as soon as the ones they know have been already mastered. This can even happen several times in a single lesson (loading a settings file is immediate).

You could say the same of Scratch. Why using Scratch instead of a full-fledged computer programming language? We are hindering their creativity by limiting what they can do by using Scratch instead of, e.g., Python or C++. This is true. But moving from PHP (or even Logo!) to Scratch increased -a lot- the audience of interested students in my courses.

> the title of this extension is _Scratch **Addons**_ which is certainly meant to increase the possibilities with Scratch.

Exactly. By adding this add-on you will extend the audience of Scratch -and of creative learning- to users that, if let by themselves, would not get interested in computer programming because they think it is "too boring", "too difficult", "too time consuming". This is the majority of my students, that have chosen to start their Communication Studies degree (not a technical degree). After my course, a lot of them gets interested in programming, but a not negligible amount of them is still just getting the minimum score on my exam, as they still think (even with all the simplifications I already applied) that this is something too vast to them. 
 
> I feel like the majority of users who use Scratch Addons wouldn't find this interesting.

I completely agree. But you are speaking of users that, interested in Scratch, are enthusiast of Scratch Addons. What I'm thinking instead it is enlarging the audience of Scratch and Scratch Addons to those users that think they will never be interested in Scratch and computer programming per se.

NOTE: I already discussed all this with the designer team of Snap. Just a few months ago, after about 10 years, they added the possibility to hide blocks from the palette. So, I'm not really in a rush. I can wait the time needed to get someone convinced that this is going to be a very useful add-on :-)

**Comment by Chiroyce1** [comment]
> After my course, a lot of them gets interested in programming, but a not negligible amount of them is still just getting the minimum score on my exam, as they still think (even with all the simplifications I already applied) that this is something too vast to them.

I see.. so this is meant to be an addon that simplifies the UI by reducing the number of blocks, makes sense. So maybe this could be a part of an addon that is similar to [Zen Mode on Visual Studio Code](https://code.visualstudio.com/docs/getstarted/userinterface#_zen-mode)? As @WorldLanguages said, this shouldn't be completely rejected, with some modifications this should be fine I suppose.

**Comment by sfederici** [comment]
Any news?

**Comment by sfederici** [comment]
I understand that #5226 can be used for storing the set of hidden blocks? I don't get what #1950 is instead.

**Comment by mxmou** [comment]
> I don't get what #1950 is instead.

It's similar, but intended for settings that can be set both on the settings page and in the addon itself.

**Comment by WorldLanguages** [comment]
For this addon, we probably won't show any settings on our settings page at all. All configuration would be done in the Scratch editor itself. This also means the Scratch Addons translators don't need to translate each of the blocks into their languages.

These settings would be synced across browsers and would also be exportable through the JSON file, just like normal settings.

In comment https://github.com/ScratchAddons/ScratchAddons/pull/6043#issuecomment-1616819356 a teacher suggests that we also provide a way to encode the ~100 yes/no choices (whether each block is shown or not) as a string of characters, so that these settings can be easily applied to multiple computers without having to download the JSON file on each one.
